src/rt/test/func/memory/memory_rc.h

```diff
@@ -0,0 +1,147 @@
+   **/
+  void test_basic()
+  {
+    // Allocate a lot of garbage.
+    {
+      auto& alloc = ThreadAlloc::get();
+      auto* o = new (alloc) C;
+
+      uintptr_t obj_count = 1;
+
+      obj_count =
+        alloc_garbage_helper<C, F, MC, MF, LC, LF, XC, XF>(alloc, o, obj_count);
+      obj_count = alloc_garbage_helper<C, C, C, XF, XF, MC, LC, LF, F, F, XC>(
+        alloc, o, obj_count);
+      obj_count = alloc_garbage_helper<C, C, C, XF, XF, MC, LC, LF, F, F, XC>(
+        alloc, o, obj_count);
+      obj_count = alloc_garbage_helper<XC, XC, XC, XF, MC>(alloc, o, obj_count);
+      obj_count =
+        alloc_garbage_helper<F, F, F, C, C, C, C, F>(alloc, o, obj_count);
+
+      Region::release(alloc, o);
+      snmalloc::debug_check_empty<snmalloc::Alloc::StateHandle>();
+    }
+
+    // Allocate a lot of objects that are all connected.
+    // Then decref the root and see if children are deallocated.
+    {
+      auto& alloc = ThreadAlloc::get();
+      auto* o = new (alloc) C;
+
+      auto* o1 = new (alloc, o) C;
+      auto* o2 = new (alloc, o) C;
+      auto* o3 = new (alloc, o) C;
+      auto* o4 = new (alloc, o) C;
+      auto* o5 = new (alloc, o) C;
+      auto* o6 = new (alloc, o) C;
+      auto* o7 = new (alloc, o) C;
+
+      // Link them up
+      o1->f1 = o2;
+      o1->f2 = o4;
+      o2->f1 = o3;
+      o2->f2 = o4;
+      RegionRc::incref(o4);
+
+      o4->f1 = o5;
+      RegionRc::incref(o5);
+      o5->f1 = o6;
+      o5->f2 = o7;
+
+      check(Region::debug_size(o) == 8);
+      check(RegionRc::debug_get_ref_count(o1) == 1);
+      check(RegionRc::debug_get_ref_count(o2) == 1);
+      check(RegionRc::debug_get_ref_count(o3) == 1);
+      check(RegionRc::debug_get_ref_count(o4) == 2);
+      check(RegionRc::debug_get_ref_count(o5) == 2);
+      check(RegionRc::debug_get_ref_count(o6) == 1);
+      check(RegionRc::debug_get_ref_count(o7) == 1);
+
+      // Decref'ing o1 to 0 should trigger a deallocation.
+      RegionRc::decref(alloc, o1, o);
+
+      check(Region::debug_size(o) == 4);
+
+      Region::release(alloc, o);
+      snmalloc::debug_check_empty<snmalloc::Alloc::StateHandle>();
+    }
+  }
+
+  /**
+   * Show that basic reference counting can't handle cycles.
+   **/
+  void test_cycles()
+  {
+    auto& alloc = ThreadAlloc::get();
+    auto* o = new (alloc) C;
+
+    // Allocate some reachable objects.
+    auto* o1 = new (alloc, o) C;
+    auto* o2 = new (alloc, o) C;
+    auto* o3 = new (alloc, o) C;
+    auto* o4 = new (alloc, o) C;
+    auto* o5 = new (alloc, o) C;
+
+    // cycle: o -> o1 -> o2 -> o3 -> o4 -> o5 -> o
+    o->f1 = o1;
+    o1->f1 = o2;
+    o2->f1 = o3;
+    o3->f1 = o4;
+    o4->f1 = o5;
+    RegionRc::incref(o);
+    o5->f1 = o;
+
+    check(Region::debug_size(o) == 6);
+    RegionRc::decref(alloc, o, o);
+    check(Region::debug_size(o) == 6);
+
+    // Break the cycle
+    RegionRc::decref(alloc, o, o);
+    check(Region::debug_size(o) == 0);
```

Comment:
Perhaps, enable this test, but comment out the assertion here saying that it will be enabled when cycle detection is enabled.